#### Group 15 - Aswin Lakshmanan Sriram & Kaviprakash Ramalingam

#### 1. Please write down the new function p(event, space) using an if clause, and the following helper function 'such_that' which applies a transformation (using a predicate argument, and a collection argument) to all possible outcomes to yield just the desired ones. A predicate is just a function that returns true or false.

In [43]:
# Function to check if a number is prime
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False 
    return True

In [44]:
D = {1,2,3,4,5,6}

D3 = {(d1, d2, d3) for d1 in D for d2 in D for d3 in D}

In [45]:
# Generate all possible outcomes
D3 = {(d1, d2, d3) for d1 in D for d2 in D for d3 in D}

In [46]:
# Function to check if the sum of the outcome is prime
def prime_sum(outcome):
    return is_prime(sum(outcome))


#### 2. Determine the probability that the sum of a three-dice roll is prime. Write a python code to compute this.

In [47]:
# predicate approach
check_predicate = callable

def p(event, space):
    if check_predicate(event):
        event = prime_that(event, space)
    return (len(event & space)/ len(space))

def such_that(predicate, collection):
    return { i for i in collection if prime_sum(i) }

print(p(prime_sum, D3))

0.33796296296296297


#### 3. What is the probability that two students in this class have the same birthday? Which one of the two probabilities, this one or the previous one (probability that someone in this class shares your birthday), do you think is higher?

In [48]:
from operator import mul
from functools import reduce

def shares_my_birthday(n):
    return n/365

def two_students_share_same_birthday(n):
    '''return 1 - 364/365 * 363/365 * .... * (365-n)/365'''
    l = range(365, 365-n, -1)
    return 1 - (reduce(mul, l,1)/(365**n))
    

print("YOU share birthday with another student: " + str(shares_my_birthday(23)))
print("2 students share birthday: " + str(two_students_share_same_birthday(23)))

YOU share birthday with another student: 0.06301369863013699
2 students share birthday: 0.5072972343239854


#### the probability that the original student (you) shares a birthday with at least one other student in a group of 23 students is approximately 6.3%. This means that, on average, when you are in a group of 23 students, there is a 6.3% chance that at least one of your classmates shares your birthday.

#### The probability that at least two students in a group of 23 students share the same birthday is approximately 50.7%. This means that, on average, when you are in a group of 23 students, there is a 50.7% chance that at least two of your classmates share the same birthday. The probability of 2 students share the same birthday with another student is higher than the remaining 23 students